docs/design/mono/embedding-api-deprecation.md

```diff
@@ -0,0 +1,68 @@
+
+We have also been forced to add a handful of .NET 5-specific API extensions, mostly to offer the Xamarin products AssemblyLoadContext-aware versions of preexisting APIs. We have defined those in the `mono-private-unstable.h` headers, but this is intended as a temporary measure to unblock some .NET 5 migrations.
+
+# Moving forward
+
+For .NET 6, we have an opportunity to clean up the existing headers and define a clear policy around ABI/API stability and deprecation.
+
+## ABI stability
+
+ABI stability is only guaranteed across major version releases. .NET 6 will not receive a servicing update that breaks ABI compatibility, but you cannot link a .NET 5 app against the .NET 6 release, full stop.
+
+## API stability
+
+API stability is more complicated. There are two distinct cases here: .NET 6 and the behavior moving forward.
+
+With the .NET 6 release specifically, we have an opportunity to do some initial API cleanup work because relatively few desktop users have migrated over to .NET 5 Mono. There is some low-hanging fruit to potentially clean up immediately, before we have significant migration: irrelevant functionality like `mono_domain_set` or `refonly` parameters, legacy compatibility hacks like `MonoLookupPInvokeStatus`, and subtly-incorrect behavior like fixed-size GC handles. 
+
+The .NET 6 immediate removal should only be used for functions that will allow for significant cleanup of the embedding API, runtime internals, or that would otherwise be likely to result in difficult to diagnose customer bugs or crashes at runtime. There are a lot of functions we might not want to remove for .NET 6 for ease of migration or simply time constraints. The earlier `mono_assembly_load` example is a potential case of this, but we have many similar cases of symbol variants. We want to remove many of them eventually, but for one of the above reasons it may not happen for .NET 6.
+
+To handle this moving forward, we will use a `MONO_DEPRECATED` define, which corresponds to the `[[deprecated]]` attribute on relevant compilers. For every API marked with this, we should have an entry on a new page on the Mono site explaining the deprecation and the suggested alternative. This creates a moderate bar for deprecations, but still leaves them as a possibility where necessary.
+
+Any function appearing in a major release marked deprecated may, and likely will go away in the following major release. This means that if `mono_assembly_load` is deprecated for .NET 6, it can be removed at any time leading up to .NET 7 or beyond, including during the previews.
+
+Deprecation is the preferred path for functions likely to see heavy use with embedders in .NET 6 and the _only_ option past that point. .NET 6 is the only release where removal may occur without the attribute first being present in a major release, though this should be avoided with functions likely to see heavy use in existing embedding workloads.
```

Comment:
It would be useful to do some customer development with embedders to validate that these assumptions are correct.